### PR TITLE
feat: complete bundle pool

### PR DESCRIPTION
**Pull request by Pythonberg1997**
### Description

add a description of your changes here...

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### CODE REVIEW

**Review by irrun — COMMENTED** [review]
also pls refine commit info

File: core/txpool/bundlepool/config.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+type Config struct {
+	Journal   string        // Journal of local transactions to survive node restarts
+	Rejournal time.Duration // Time interval to regenerate the local transaction journal
+
+	PriceLimit uint64 // Minimum gas price to enforce for acceptance into the pool
+	PriceBump  uint64 // Minimum price bump percentage to replace an already existing transaction (nonce)
+
+	BundleSlots     uint64 // Maximum number of bundle slots for all accounts
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply core/txpool/bundlepool/config.go:16]: Done

File: core/txpool/bundlepool/bundlepool.go [hunk 2]
```diff
@@ -49,57 +52,115 @@ type BundleSimulator interface {
...
-	PriceBump  uint64 // Minimum price bump percentage to replace an already existing transaction (nonce)
-
-	BundleSlot      uint64 // Maximum number of bundle slots for all accounts
-	GlobalQueue     uint64 // Maximum number of non-executable bundle slots for all accounts
-	MaxBundleBlocks uint64 // Maximum number of blocks for calculating MinimalBundleGasPrice
+type BundlePool struct {
+	config  Config
+	chain   BlockChain
+	mu      sync.RWMutex
+	reserve txpool.AddressReserver // Address reserver to ensure exclusivity across subpools
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply core/txpool/bundlepool/bundlepool.go:59]: Done

File: core/txpool/bundlepool/bundlepool.go [hunk 3]
```diff
@@ -109,188 +170,191 @@ func (pool *BundlePool) AddBundle(bundle *types.Bundle) error {
 	hash := crypto.Keccak256Hash(bz)
 	bundle.Hash = hash
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply core/txpool/bundlepool/bundlepool.go:171]: Done

File: core/txpool/bundlepool/bundlepool.go [hunk 4]
```diff
@@ -49,57 +52,115 @@ type BundleSimulator interface {
...
-	BundleGasPricerExpireTime     time.Duration // Store time duration amount of recent mev gas price
-	UpdateBundleGasPricerInterval time.Duration // Time interval to update MevGasPricePool
-}
+	gasTip *uint256.Int // Currently accepted minimum gas tip
 
-type BundlePool struct {
-	config Config
-	chain  BlockChain
-	mu     sync.RWMutex
+	bundles map[common.Hash]*types.Bundle
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply core/txpool/bundlepool/bundlepool.go:63]: Done

File: core/txpool/bundlepool/bundlepool.go [hunk 5]
```diff
@@ -109,188 +170,191 @@ func (pool *BundlePool) AddBundle(bundle *types.Bundle) error {
...
+func (p *BundlePool) PendingBundles(blockNumber *big.Int, blockTimestamp uint64) []*types.Bundle {
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
 	// returned values
 	var ret []*types.Bundle
-	// rolled over values
-	bundles := make(map[common.Hash]*types.Bundle)
+	for hash, bundle := range p.bundles {
+		if bundle.Txs.Len() == 0 {
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply core/txpool/bundlepool/bundlepool.go:221]: Removed

File: core/txpool/bundlepool/bundlepool.go [hunk 6]
```diff
@@ -109,188 +170,191 @@ func (pool *BundlePool) AddBundle(bundle *types.Bundle) error {
...
+			p.PruneBundle(hash)
+			continue
+		}
 
-	for uid, bundle := range pool.bundles {
 		// Prune outdated bundles
 		if (bundle.MaxTimestamp != 0 && blockTimestamp > bundle.MaxTimestamp) ||
 			blockNumber.Cmp(big.NewInt(bundle.MaxBlockNumber)) > 0 {
+			p.PruneBundle(hash)
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply core/txpool/bundlepool/bundlepool.go:229]: Fixed

**Review by irrun — APPROVED** [review]

**Pull request merged**